VIMC-4734 Add comments to responsibilities

mwoodbri opened the pull request:
Summary:
* Add new endpoint that returns any comments attached to responsibilities
* This is a new endpoint as it is permissioned differently to the existing responsibilities endpoint. However, it returns data structured so that it is easy to map between the two.
* Requires vimc/montagu-db#159 and vimc/montagu-webmodels#96

To-do:
- [x] Documentation on endpoint
- [x] Updates to `Security.md` and `spec.md`
- [x] Update config/submodules to point at `master`

EmmaLRussell commented on src/app/src/main/kotlin/org/vaccineimpact/api/app/repositories/jooq/JooqResponsibilitiesRepository.kt:
```diff
@@ -185,6 +188,46 @@ class JooqResponsibilitiesRepository(
+                                .and(Tables.RESPONSIBILITY.IS_OPEN)
+                                .fetch()
+                                .intoMap(Tables.SCENARIO_DESCRIPTION.ID)
+                        val responsibilities = scenarioRepository.getScenarios(records.keys).map { scenario ->
```
Is there a danger this get will get scenarios from out of touchstone if just filtering on the scenario description id? Scenario description not touchstone-specific (but scenario is). 

Other than that, I think the logic here looks right! I think my tendency would have been to try to reduce the number of queries by doing one big query to get the responsibilities and then grouped by responsibility set id, then done another query to get the latest comments. But this is probably more readable, and if performance isn't an issue no reason to limit number of queries.

mwoodbri replied:
> Other than that, I think the logic here looks right! I think my tendency would have been to try to reduce the number of queries by doing one big query to get the responsibilities and then grouped by responsibility set id, then done another query to get the latest comments.

Done - comments on this change in particular very welcome

EmmaLRussell commented on src/app/src/test/kotlin/org/vaccineimpact/api/tests/controllers/TouchstoneControllerTests.kt:
```diff
@@ -181,10 +183,56 @@ class TouchstoneControllerTests : MontaguTests()
+            on { touchstoneVersions } doReturn InMemoryDataSet(listOf(openTouchstoneVersion))
+        }
+        val responsibilityRepo = mock<ResponsibilitiesRepository> {
+            on { it.getResponsibilitiesWithCommentsForTouchstone(any()) } doReturn sets
```
Don't think you need to mock this for this test?

EmmaLRussell reviewed (commented):
Looks good to me, couple of queries. 

Should there be a models submodule PR too?

Could do with an integration test aka blackbox test?

mwoodbri commented on src/app/src/main/kotlin/org/vaccineimpact/api/app/controllers/TouchstoneController.kt:
```diff
@@ -74,6 +80,29 @@ class TouchstoneController(
+    {
+        val touchstoneVersion = touchstoneVersion(context, touchstoneRepo)
+        val groupId = context.queryParams("group_id")!!
+        val scenarioDescriptionId: String = context.queryParams("scenario_id")!!
```
Good question. It does. Slightly confusingly the [`Scenario`](https://github.com/vimc/montagu-webmodels/blob/master/src/main/kotlin/org/vaccineimpact/api/models/Scenario.kt) class maps to the [`scenario_description`](https://www.vaccineimpact.org/montagu-db-docs/latest/tables/scenario_description.html) table, and it's here that the gap is bridged.

mwoodbri commented on src/app/src/main/kotlin/org/vaccineimpact/api/app/app_start/route_config/TouchstoneRouteConfig.kt:
```diff
@@ -23,6 +24,15 @@ object TouchstoneRouteConfig : RouteConfig
+                    .json()
+                    .secure(responsibilityReviewPermissions),
+
+            Endpoint("$baseUrl:touchstone-version-id/responsibilities/comments/", controller, "addResponsibilityComment")
```
I'm open-minded but I think the case for `addResponsibilityComment` being here is that:
- comments are being attached in the context of a particular touchstone
- keeping the endpoints together feels more maintainable
- they probably both need the same permissions, which are different to those currently applied in `ResponsibilityRouteConfig`

Regarding the route itself, I agree with the comment text should be in the post body as JSON for consistency with other endpoints. For the URL we have the following options:
1. `:touchstone-version-id/responsibilities/comments/` with body of `{comment, group_id, scenario_id}`
2. `:touchstone-version-id/responsibilities/comments/:group-id/:scenario-id/ with body of `{comment}`
3. `:touchstone-version-id/responsibilities/:group-id/:scenario-id/comments/ with body of `{comment}`

There is an argument for each... I would favour option 1 if starting from scratch, but the other two are more consistent with other endpoints. Option 2 would make more sense for a GET, as it would imply returning a subset of comments. Option 3 is more consistent with coverage upload URL structure so I'm inclined to that one. Thoughts welcome!

EmmaLRussell replied:
Fair point on the route_config location. 

Yeah, 3 is preferable out of that list I'd say. To be consistent with [getRespnsibility](https://github.com/vimc/montagu-api/blob/master/src/app/src/main/kotlin/org/vaccineimpact/api/app/app_start/route_config/ResponsibilityRouteConfig.kt#L27) it would probably be `/modelling-groups/:group-id/responsibilities/:touchstone-version-id/:scenario-id/comments` - but then that would perhaps make less sense in the context of keeping the route in `TouchstoneRouteConfig`

mwoodbri commented:
Changes since last review:
* Documentation for new endpoints, including schemas
* Change POST URL
* Simplify SQL used to retrieve comments
* Make controller test more thorough
* Add black-box tests
* Minor fix to instructions for running black-box tests

EmmaLRussell reviewed (approved):
This looks great to me, pending the change to the model classes discussed in https://github.com/vimc/montagu-webmodels/pull/96
